### PR TITLE
Added multiple initializations and parallelization

### PR DESCRIPTION
**Pull request by holgerteichgraeber**
The following features were added:
- multiple initializations of the k-shape algorithm. Automatically picks the best centroid (lowest SBD) amongst all initializations. 
- maximum number of iterations is a parameter (default max_iter=100)
- Parallelization for the multiple initializations is implemented similar to the parallelization of k-means clustering in scikit-learn.

### CODE REVIEW

File: kshape/core.py [hunk 1]
```diff
@@ -155,34 +157,83 @@ def _extract_shape(idx, x, j, cur_center):
...
+def _kshape(x, k,n_init=1, max_iter=100, n_jobs = 1, random_state=None):
     """
     >>> from numpy.random import seed; seed(0)
     >>> _kshape(np.array([[1,2,3,4], [0,1,2,3], [-1,1,-1,1], [1,2,2,3]]), 2)
     (array([0, 0, 1, 0]), array([[-1.2244258 , -0.35015476,  0.52411628,  1.05046429],
            [-0.8660254 ,  0.8660254 , -0.8660254 ,  0.8660254 ]]))
     """
-    m = x.shape[0]
-    idx = randint(0, k, size=m)
+    random_state = check_random_state(random_state)
```
Comment (Mic92) [inline kshape/core.py]: So this creates a different random clustering every time it runs? Would a fixed seed approach not be better?

File: kshape/core.py [hunk 2]
```diff
@@ -155,34 +157,83 @@ def _extract_shape(idx, x, j, cur_center):
...
+    best_tot_dist,best_centroids,best_idx = None,None,None
+    
+    if n_jobs ==1:
+
+        for i_init in range(n_init):    
+            # n_init is the number of random starting points
+            # pdb.set_trace()
```
Comment (Mic92) [inline kshape/core.py]: Should be removed.

File: kshape/core.py [hunk 3]
```diff
@@ -155,34 +157,83 @@ def _extract_shape(idx, x, j, cur_center):
...
+            
+            idx, centroids,tot_dist, iterations = _kshape_single(x, k, max_iter=max_iter, random_state= random_state) 
+            if best_tot_dist is None or tot_dist < best_tot_dist:
+                best_idx = idx.copy()
+                best_centroids = centroids.copy()
+                best_tot_dist = tot_dist
+    else: # n_jobs not =1 # if -1, all CPUs are used
```
Comment (Mic92) [inline kshape/core.py]: The semantics of `n_jobs` should be not hidden in comments, but publicly documented.

File: kshape/core.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
 from numpy.random import randint
 from numpy.linalg import norm, eigh
 from numpy.fft import fft, ifft
+from sklearn.externals.joblib import Parallel,delayed
+from sklearn.utils import check_random_state
```
Comment (Mic92) [inline kshape/core.py:8]: This is a new dependency and should be documented/added to https://github.com/Mic92/kshape/blob/master/setup.py#L37

Comment (Mic92) [reply kshape/core.py:8]: It might be desirable to make it optional as it is not supported by pypy yet as for the not parallel version only `check_random_state` is required, which is just a wrapper around numpy random: https://github.com/scikit-learn/scikit-learn/blob/f0ab589f/sklearn/utils/validation.py#L617

**Review by Mic92 — COMMENTED** [review]
I have not worked with multiple initialization points. Does kshape not eventually converge to the same clustering?

File: kshape/core.py (hunk 4 above)
Comment (holgerteichgraeber) [reply kshape/core.py:8]: Added to setup

How could we make it optional?

File: kshape/core.py (hunk 1 above)
Comment (holgerteichgraeber) [reply kshape/core.py]: The partitional clustering algorithm only converges to local solutions. Even though each initialization eventually converges, it converges to a different solution. 

If you fix the seed before calling kshape(), it will run the same clustering every time you run kshape(). See https://github.com/holgerteichgraeber/kshape/blob/master/demo/example_fixedseed.py

Comment (Mic92) [reply kshape/core.py]: Yes, but then loop would create different starting states in a deterministic manner, no? Having a fixed seed by default makes unit testing much easier.

Comment (Mic92) [reply kshape/core.py]: I suppose equal distribution is the goal here, not true randomness.

File: kshape/core.py (hunk 4 above)
Comment (Mic92) [reply kshape/core.py:8]: I don't think scikit is required at all. It just wraps `np.random.RandomState` and [joblib](https://github.com/joblib/joblib). Does this code benefit from `joblib`? Otherwise [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) might be sufficient.

File: kshape/core.py [hunk 5]
```diff
@@ -158,6 +158,11 @@ def _extract_shape(idx, x, j, cur_center):
 
 
 def _kshape(x, k,n_init=1, max_iter=100, n_jobs = 1, random_state=None):
+    """
+    Runs multiple initializations of kshape and returns the best in terms of ShapeBasedDistance. 
+    max_iter is the maximum number of iterations within one run, 
+    n_jobs is the number of jobs: 1 is non-parallel version, if -1, all CPUs are used
+    """
```
Comment (Mic92) [inline kshape/core.py]: `_kshape` is internal, I think you wanted to add it to `kshape()` instead.

File: kshape/core.py (hunk 4 above)
Comment (holgerteichgraeber) [reply kshape/core.py:8]: I use the helper for embarrassingly parallel code from joblib.

File: kshape/core.py (hunk 1 above)
Comment (holgerteichgraeber) [reply kshape/core.py]: Yes, it would create different starting states in a deterministic manner. I'm not sure if I follow why that is an issue. Why is this different from having a fixed seed by default?

File: kshape/core.py (hunk 4 above)
Comment (holgerteichgraeber) [reply kshape/core.py:8]: Got rid of the scikit reference for joblib. Will get rid of the scikit reference for random_state once we figure what to do there.

File: kshape/core.py (hunk 5 above)
Comment (holgerteichgraeber) [reply kshape/core.py]: Good call. Fixed in 
https://github.com/Mic92/kshape/pull/8/commits/3427048d8b0d985b706650eebc1597ac09339c9 .
